### PR TITLE
`partial=`

### PR DESCRIPTION
**Pull request by camsaul**
New invention.

Adds a new test assertion type called `partial=` that ignores extra keys/elements in `actual`.  It's a little annoying that we have to go around updating hundreds of tests every time we add a new column to a table, and 99% of the time we don't care what the value of that column is anyway. No more. Now we can just assert that things we want are in there and ignore other stuff that gets added in the future.

```clj
(t/deftest is-cam-test
  (t/is (partial= {:id   100
                   :name "Cam"}
                  {:id     100
                   :name   "Cam"
                   :food   "Ham"
                   :animal "Ram"})))
;; -> ok

(t/deftest is-cam-test
  (t/is (partial= {:id   100
                   :name "Cam"}
                  {:id     100
                   :name   "Sam"
                   :food   "Ham"
                   :animal "Ram"})))
;; =>
;;
;; Fail in is-cam-test
;; 
;; expected: {:id 100, :name "Cam"}
;; 
;;   actual: {:id 100, :name "Sam", :food "Ham", :animail "Ram"}          
;;     diff: - {:name "Cam"}          
;;           + {:name "Sam"}            
```

### More examples:

```clj
(partial= {:a 1}
          {:a 1})
;; => :pass

(partial= {:a 1}
          {:a 1, :b 2})
;; => :pass          

(partial= {:a 1}
          {:a 2})
;; =>
;;
;; Fail in some-test
;; 
;; expected: {:a 1}
;; 
;;   actual: {:a 2}          
;;     diff: - {:a 1}          
;;           + {:a 2}            

(partial= {:a 1}
          {:b 2})
;; =>
;;
;; Fail in some-test
;; 
;; expected: {:a 1}            
;; 
;;   actual: {:b 2}          
;;     diff: - {:a 1}          
;;           + {}            

(partial= {:a 1}
          {:a 2, :b 1})
;; =>
;;
;; Fail in some-test
;; 
;; expected: {:a 1}            
;; 
;;   actual: {:a 2, :b 1}          
;;     diff: - {:a 1}          
;;           + {:a 2}

(partial= [1 2 {:x 3}]
          [1 2 {:x 3}])
;; => :pass

(partial= [1 2 {:x 3}]
          [1 2])
;; =>
;;
;; Fail in some-test
;; 
;; expected: [1 2 {:x 3}]            
;; 
;;   actual: [1 2]          
;;     diff: - [nil nil {:x 3}]          
;;           + nil            

(partial= [1 2 {:x 3}]
          nil)
;; =>
;;
;; Fail in some-test
;; 
;; expected: [1 2 {:x 3}]            
;; 
;;   actual: nil          
;;     diff: - [1 2 {:x 3}]          
;;           + nil            

(partial= [1 2 {:x 3}]
          [1 2 {:x 1}])
;; =>
;;
;; Fail in some-test
;; 
;; expected: [1 2 {:x 3}]            
;; 
;;   actual: [1 2 {:x 1}]          
;;     diff: - [nil nil {:x 3}]          
;;           + [nil nil {:x 1}]            

(partial= [1 2 {:x 3}]
          [1 2 3])
;; =>
;;
;; Fail in some-test
;; 
;; expected: [1 2 {:x 3}]            
;; 
;;   actual: [1 2 3]          
;;     diff: - [nil nil {:x 3}]          
;;           + [nil nil 3]            

(partial= [1 2 {:x 3}]
          [1 2 {:x 3} 4])
;; => :pass

(partial= [1 2 {:x 3}]
          [1 2 {:x 2} 4])
;; =>
;;
;; Fail in some-test
;; 
;; expected: [1 2 {:x 3}]            
;; 
;;   actual: [1 2 {:x 2} 4]          
;;     diff: - [nil nil {:x 3}]          
;;           + [nil nil {:x 2}]            

(partial= [1 2 {:x 3}]
          [1 2 {:x 1, :y 2} 4])
;; =>
;;
;; Fail in some-test
;; 
;; expected: [1 2 {:x 3}]            
;; 
;;   actual: [1 2 {:x 1, :y 2} 4]          
;;     diff: - [nil nil {:x 3}]          
;;           + [nil nil {:x 1}]

(partial= [:field 100]
          [:field 100 {::abc "wow"}])
;; => :pass          
```

### CODE REVIEW

**Comment by jeff303** [comment]
Seems like a more useful version of `metabase.test.util/select-keys-sequentially`, which I added at some point to address a similar need.

File: test/metabase/test_runner/effects.clj [hunk 1]
```diff
@@ -68,3 +68,46 @@
...
   `(t/do-report
     (query=-report ~message ~expected ~actual)))
+
+;; `partial=` is like `=` but only compares stuff (using [[data/diff]] that's in `expected`. Anything else is ignored.
+
+(defn- remove-keys-not-in-expected
+  "Remove all the extra stuff (i.e. extra map keys or extra sequence elements) from the `actual` diff that's not
+  in the original `expected` form."
+  [expected actual]
+  (println "(pr-str expected) (pr-str actual):" (pr-str expected) (pr-str actual)) ; NOCOMMIT
```
Comment (howonlee) [inline test/metabase/test_runner/effects.clj:78]: looks like the nocommit didn't work?

Comment (camsaul) [inline test/metabase/test_runner/effects.clj:78]:
```suggestion
```

**Review by dpsutton — APPROVED** [review]
Approving, but this could get really difficult to reason about with sequential things.

```clojure
(t/deftest foo (t/is (partial= [{:foo :bar} {:baz :quz}]
                               [{:baz :quz}])))

(foo)

expected: [{:foo :bar} {:baz :quz}]
  actual: [{:baz :quz}]
    diff: - [{:foo :bar} {:baz :quz}]
          + [nil]

```

The `actual` stays true but the diffs will be murder. I think most of our situations will have homogeneous collections more or less, but something to remember if diffs ever seem unhelpful.

**Comment by camsaul** [comment]
Docker failed because of the branch name but besides that this PR is green so going to go ahead and merge

**Pull request merged**